### DISCUSSION TITLE
Directly compile cplusplus code into an application VS. compile cplusplus code as a library then linked to an application

### DISCUSSION THREAD

**Discussion by huskier** [libvips/libvips#4271] [Q&A]
When we directly compile cplusplus code into an application, the application can work properly.
When we compile cplusplus code as a library then linked to an application, the application cannot be generated because of the error of "unresolved external symbol errors".

The code to compile cplusplus code as a library is as following. It is a QT pro file. We could get vipscplusplus.lib file, but when we link the library into our application, the above errors (unresolved external symbol errors) occur.

```
TARGET = vipscplusplus
TEMPLATE = lib
VERSION = 0.1.0
CONFIG += staticlib c++17
DESTDIR = $$PWD/../../../lib
INCLUDEPATH += C:\Application\Qt\5.15.2\msvc2019_64\include
INCLUDEPATH += C:\Application\Qt\5.15.2\msvc2019_64\include\vips
INCLUDEPATH += C:\Application\Qt\5.15.2\msvc2019_64\include\vips\glib-2.0
SOURCES += \
        VConnection.cpp \
        VError.cpp \
        VImage.cpp \
        VInterpolate.cpp \
        vips-operators.cpp \
        VRegion.cpp
```

The error text looks like the following:

`vipscplusplus.lib(VImage.obj):-1: error: LNK2019: unresolved external symbol "__declspec(dllimport) public: static class vips::VRegion __cdecl vips::VRegion::new_from_image(class vips::VImage)" (__imp_?new_from_image@VRegion@vips@@SA?AV12@VVImage@2@@Z)，function "public: class vips::VRegion __cdecl vips::VImage::region(void)const " (?region@VImage@vips@@QEBA?AVVRegion@2@XZ)  called this symbol.`

We're on Windows 11 with MSVC2019 64bit compiling tool chain and QT 5.15.2 community version.

**Comment by kleisauke** [comment]
Could you try setting `_VIPS_PUBLIC` to `__declspec(dllexport)`?:
https://github.com/libvips/libvips/blob/e38953c5dfa302875c4578ad1222b114ee6eac69/cplusplus/include/vips/vips8#L43-L48

In `qmake`, this can be configured as follows:
```
DEFINES += _VIPS_PUBLIC=__declspec(dllexport)
```

**Comment by huskier** [comment]
This is the key to success. Thank you very much, @kleisauke.
> DEFINES += _VIPS_PUBLIC=__declspec(dllexport)

The QT pro file now is :

```
TARGET = vipscplusplus
TEMPLATE = lib
VERSION = 0.1.0
CONFIG += staticlib c++17
DESTDIR = $$PWD/../../../lib
DEFINES += _VIPS_PUBLIC=__declspec(dllexport)
INCLUDEPATH += C:\Application\Qt\5.15.2\msvc2019_64\include
INCLUDEPATH += C:\Application\Qt\5.15.2\msvc2019_64\include\vips
INCLUDEPATH += C:\Application\Qt\5.15.2\msvc2019_64\include\vips\glib-2.0
SOURCES += \
        VConnection.cpp \
        VError.cpp \
        VImage.cpp \
        VInterpolate.cpp \
        vips-operators.cpp \
        VRegion.cpp

```
